google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/OpenTelemetryUtil.java

```diff
@@ -0,0 +1,198 @@
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.util.List;
+
+public class OpenTelemetryUtil {
+  private static final String MESSAGING_SYSTEM_VALUE = "gcp_pubsub";
+  private static final String PROJECT_ATTR_KEY = "gcp.project_id";
+  private static final String ACK_DEADLINE_ATTR_KEY = "messaging.gcp_pubsub.message.ack_deadline";
+  private static final String RECEIPT_MODACK_ATTR_KEY = "messaging.gcp_pubsub.is_receipt_modack";
+
+  private static final String PUBLISH_RPC_SPAN_SUFFIX = " publish";
+
+  /** Populates attributes that are common the publisher parent span and publish RPC span. */
+  public static final AttributesBuilder createCommonSpanAttributesBuilder(
+      String destinationName, String projectName, String codeFunction, String operation) {
+    AttributesBuilder attributesBuilder =
+        Attributes.builder()
+            .put(SemanticAttributes.MESSAGING_SYSTEM, MESSAGING_SYSTEM_VALUE)
+            .put(SemanticAttributes.MESSAGING_DESTINATION_NAME, destinationName)
+            .put(PROJECT_ATTR_KEY, projectName)
+            .put(SemanticAttributes.CODE_FUNCTION, codeFunction);
+    if (operation != null) {
+      attributesBuilder.put(SemanticAttributes.MESSAGING_OPERATION, operation);
+    }
+
+    return attributesBuilder;
+  }
+
+  /**
+   * Creates, starts, and returns a publish RPC span for the given message batch. Bi-directional
+   * links with the publisher parent span are created for sampled messages in the batch.
+   */
+  public static final Span startPublishRpcSpan(
+      Tracer tracer,
+      String topic,
+      List<PubsubMessageWrapper> messages,
+      boolean enableOpenTelemetryTracing) {
+    if (enableOpenTelemetryTracing && tracer != null) {
+      TopicName topicName = TopicName.parse(topic);
+      Attributes attributes =
+          createCommonSpanAttributesBuilder(
+                  topicName.getTopic(), topicName.getProject(), "Publisher.publishCall", "publish")
+              .put(SemanticAttributes.MESSAGING_BATCH_MESSAGE_COUNT, messages.size())
+              .build();
+      Span publishRpcSpan =
+          tracer
+              .spanBuilder(topicName.getTopic() + PUBLISH_RPC_SPAN_SUFFIX)
+              .setSpanKind(SpanKind.CLIENT)
+              .setAllAttributes(attributes)
+              .startSpan();
+
+      for (PubsubMessageWrapper message : messages) {
+        if (message.getPublisherSpan().getSpanContext().isSampled()) {
+          Attributes linkAttributes =
+              Attributes.builder().put(SemanticAttributes.MESSAGING_OPERATION, "publish").build();
+          publishRpcSpan.addLink(message.getPublisherSpan().getSpanContext(), linkAttributes);
```

Comment:
Can you do addLink earlier rather than here? It might make the code a bit more complicated, but it's preferred to do so as part of the span builder: https://javadoc.io/doc/io.opentelemetry/opentelemetry-api/latest/io/opentelemetry/api/trace/Span.html#addLink(io.opentelemetry.api.trace.SpanContext)

This is due to sampling decisions not respecting addLink called after span creation.

Reply:
I've moved the `addLink` calls for the `publishRpcSpan` to occur before the span is started.